src/index.ts

```diff
@@ -110,37 +80,86 @@ export function lambdaHandler<TContext extends BaseContext>(
         },
         body: httpGraphQLResponse.completeBody,
       };
-    } catch (error) {
-      throw error;
+    } catch (e) {
+      return {
+        statusCode: 400,
+        body: (e as Error).message,
+      };
     }
   };
 }
 
-function createGraphQLRequest(
-  parsedBody: string | object | undefined,
-  event: APIGatewayProxyEvent | APIGatewayProxyEventV2,
-): HTTPGraphQLRequest {
-  const headers = new Map<string, string>();
-  for (const [key, value] of Object.entries(event.headers)) {
-    if (value !== undefined) {
-      // Node/Express headers can be an array or a single value. We join
-      // multi-valued headers with `, ` just like the Fetch API's `Headers`
-      // does. We assume that keys are already lower-cased (as per the Node
-      // docs on IncomingMessage.headers) and so we don't bother to lower-case
-      // them or combine across multiple keys that would lower-case to the
-      // same value.
-      headers.set(key, Array.isArray(value) ? value.join(', ') : value);
-    }
+function normalizeGatewayEvent(event: GatewayEvent): HTTPGraphQLRequest {
+  if (isV1Event(event)) {
+    return normalizeV1Event(event);
+  } else if (isV2Event(event)) {
+    return normalizeV2Event(event);
+  } else {
+    throw Error('Unknown event type');
   }
+}
+
+function isV1Event(event: GatewayEvent): event is APIGatewayProxyEvent {
+  return !('version' in event);
+}
+
+function isV2Event(event: GatewayEvent): event is APIGatewayProxyEventV2 {
+  return 'version' in event && event.version === '2.0';
+}
+
+function normalizeV1Event(event: APIGatewayProxyEvent): HTTPGraphQLRequest {
+  const headers = normalizeHeaders(event.headers);
+  const body = parseBody(event.body, headers.get('content-type'));
+  const reconstructedSearchString = Object.entries({
+    ...event.queryStringParameters,
+    ...event.multiValueQueryStringParameters,
+  }).reduce((queryString, next, i) => {
+    const delimiter = i === 0 ? '?' : '&';
+    const addition = Array.isArray(next[1])
+      ? next[1].map((v) => `${next[0]}=${v}`).join('&')
+      : `${next[0]}=${next[1]}`;
+    return `${queryString}${delimiter}${addition}`;
+  }, '');
```

Comment:
Would it be a lot cleaner and consistent to use `URLSearchParams`?

```suggestion
  // Single value parameters can be directly added
  const searchParams = new URLSearchParams(event.queryStringParameters);
  // Passing a key with an array entry to the constructor yields
  // one value in the querystring with %2C as the array was flattened to a string
  // Multi values must be appended individually to get the to-spec output
  for (const [key, values] of Object.entries(event.multiValueQueryStringParameters)) {
    for(const value of values) {
      searchParams.append(key, value);
    }
  }
  const searchParamString = `?${searchParams.toString()}`
```

Reply:
This is a great suggestion, much happier using a builtin to handle this.

TypeScript needed a little convincing w.r.t. some `undefined` handling so my solution is close to yours but needed a little extra. Lmk what you think.

[e2c9e7f](https://github.com/michael-watson/apollo-server-integration-aws-lambda/pull/6/commits/e2c9e7f5cdc6cdd0703e44bb175ca9535543b354)

Reply:
Looks good! I just wrote that in the github suggestion editor so I figured something would need changing. Your implementation looks great!